src/hooks/gatewayAuthToken/useGatewayAuthToken.ts

```diff
@@ -12,20 +12,45 @@ import {
     getGatewayAuthTokenSettings,
     getSupabaseAnonymousKey,
 } from 'utils/env-utils';
-import {
-    getStoredGatewayAuthConfig,
-    storeGatewayAuthConfig,
-} from 'utils/localStorage-utils';
+import { getStoredGatewayAuthConfig, storeGatewayAuthConfig } from './cache';
 
 const { gatewayAuthTokenEndpoint } = getGatewayAuthTokenSettings();
 
+interface GatewayAuthTokenFetcherResponse extends GatewayAuthTokenResponse {
+    cached?: boolean;
+}
+
 // The request body for this API is a string array corresponding to the prefixes a user has access to.
-type GatewayFetcherArgs = [string, string[], string | undefined];
-export const gatewayFetcher = ({
+type GatewayFetcherArgs = [string, string[], string | undefined, string];
+const gatewayFetcher = ({
     0: endpoint,
     1: prefixes,
     2: sessionKey,
-}: GatewayFetcherArgs): Promise<GatewayAuthTokenResponse[]> => {
+    3: cacheKey,
+}: GatewayFetcherArgs): Promise<GatewayAuthTokenFetcherResponse[]> => {
+    // Grab the Gateway token from local storage
+    let jwt: JWTPayload | undefined,
+        gatewayConfig: GatewayAuthTokenResponse | undefined;
+    try {
+        gatewayConfig = getStoredGatewayAuthConfig(cacheKey);
+        jwt = gatewayConfig ? decodeJwt(gatewayConfig.token) : undefined;
+    } catch {
+        jwt = undefined;
+        gatewayConfig = undefined;
+    }
+
+    // Check if the Gateway token is expired and we need to get a new one
+    let tokenExpired = true;
+    if (jwt?.exp) {
+        tokenExpired = isBefore(jwt.exp * 1000, Date.now());
+    }
+
+    // If the token is still good and it contains a URL (mainly checking b/c typescript) then just
+    //  used the cached value.
+    if (!tokenExpired && gatewayConfig?.gateway_url) {
+        return Promise.resolve([{ ...gatewayConfig, cached: true }]);
+    }
```

Comment:
Down below we pretty much always end up calling the swr call to fetch this. So keeping the caching approach in the fetcher made it easier to do.